app/services/subscription_variants_service.rb

```diff
@@ -0,0 +1,39 @@
+class SubscriptionVariantsService
+  # Includes the following variants:
+  # - Variants of permitted producers
+  # - Variants of hub
+  # - Variants that are in outgoing exchanges where the hub is receiver
+  def self.eligible_variants(distributor)
+    variant_conditions = ["spree_products.supplier_id IN (?)", permitted_producer_ids(distributor)]
+    exchange_variant_ids = outgoing_exchange_variant_ids(distributor)
+    if exchange_variant_ids.present?
+      variant_conditions[0] << " OR spree_variants.id IN (?)"
+      variant_conditions << exchange_variant_ids
+    end
+
+    Spree::Variant.joins(:product).where(is_master: false).where(*variant_conditions)
+  end
+
+  def self.in_open_and_upcoming_order_cycles?(distributor, schedule, variant)
+    scope = ExchangeVariant.joins(exchange: { order_cycle: :schedules })
+      .where(variant_id: variant, exchanges: { incoming: false, receiver_id: distributor })
+      .merge(OrderCycle.not_closed)
+    scope = scope.where(schedules: { id: schedule })
```

Comment:
why don't we simply chain together this last `where`? We're triggering two db queries while I think one would do.

Reply:
@sauloperez Not sure if I understand your concern correctly... This wouldn't actually trigger two DB queries but just one: setting up the scope does not immediately execute the query on the DB. Except e.g. if running these lines in the console, because the console would then run `inspect` (or similar) on the result of each line.

Reply:
That's right, sorry. Still, I wonder why we need to build the relation in two steps. Is it for readability? just for me to know.

Reply:
@sauloperez Yes, for readability. Although it could have been even better if I extracted the first line into another method.